[SPARK-20732][CORE] Decommission cache blocks to other executors when an executor is decommissioned

prakharjain09 opened the pull request:
### What changes were proposed in this pull request?
After changes in SPARK-20628, CoarseGrainedSchedulerBackend can decommission an executor and stop assigning new tasks on it. We should also decommission the corresponding blockmanagers in the same way. i.e. Move the cached RDD blocks from those executors to other active executors.

### Why are the changes needed?
We need to gracefully decommission the block managers so that the underlying RDD cache blocks are not lost in case the executors are taken away forcefully after some timeout (because of spotloss/pre-emptible VM etc). Its good to save as much cache data as possible. 

Also In future once the decommissioning signal comes from Cluster Manager (say YARN/Mesos etc), dynamic allocation + this change gives us opportunity to downscale the executors faster by making the executors free of cache data.

Note that this is a best effort approach. We try to move cache blocks from decommissioning executors to active executors. If the active executors don't have free resources available on them for caching, then the decommissioning executors will keep the cache block which it was not able to move and it will still be able to serve them.

Current overall Flow:

1. CoarseGrainedSchedulerBackend receives a signal to decommissionExecutor. On receiving the signal, it do 2 things - Stop assigning new tasks (SPARK-20628), Send another message to BlockManagerMasterEndpoint (via BlockManagerMaster) to decommission the corresponding BlockManager.

2. BlockManagerMasterEndpoint receives "DecommissionBlockManagers" message. On receiving this, it moves the corresponding block managers to "decommissioning" state. All decommissioning BMs are excluded from the getPeers RPC call which is used for replication. All these decommissioning BMs are also sent message from BlockManagerMasterEndpoint to start decommissioning process on themselves.

3. BlockManager on worker (say BM-x) receives the "DecommissionBlockManager" message. Now it will start BlockManagerDecommissionManager thread to offload all the RDD cached blocks. This thread can make multiple reattempts to decommission the existing cache blocks (multiple reattempts might be needed as there might not be sufficient space in other active BMs initially).

### Does this PR introduce any user-facing change?
NO

### How was this patch tested?
Added UTs.

prakharjain09 committed:
[SPARK-20732][CORE] Decommission cache blocks to other executors when an executor is decommissioned

prakharjain09 committed:
changes

prakharjain09 commented:
cc - @cloud-fan @holdenk

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented:
Jenkins ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24352/
Test PASSed.

SparkQA commented:
**[Test build #119623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119623/testReport)** for PR 27864 at commit [`6a47615`](https://github.com/apache/spark/commit/6a47615223e5b1454811da68fa5f52e58d3b8076).

dongjoon-hyun commented:
Thank you for making a PR, @prakharjain09 .

SparkQA commented:
**[Test build #119623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119623/testReport)** for PR 27864 at commit [`6a47615`](https://github.com/apache/spark/commit/6a47615223e5b1454811da68fa5f52e58d3b8076).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/119623/
Test PASSed.

prakharjain09 commented:
@holdenk @cloud-fan  This PR currently does not handle running tasks - What to do with tasks that are already running on these decommissioned executors? If we start BlockManager decommissioning and we stop accepting new cached RDD blocks, then these already running tasks might fail (In case these already running tasks try to cache something).

One possible approach is: Kill these already running tasks after some configurable timeout.

Executor decommissioning can be broken into two steps:
1. Compute decommissioning: Stop assigning new tasks, Kill existing running tasks - maybe after some timeout
2. Storage decommissioning: BlockManager stops taking new cache blocks, offloads old  cache blocks

Firstly we can do `Compute decommissioning` i.e. Stop assigning new tasks (already done as part of  SPARK-20628), kill running tasks after some timeout (TODO)
Then do `Storage decommissioning` i.e. Stop accepting new RDD cache blocks (done as part of this PR), Move existing RDD cache blocks to other possible locations (done as part of this PR), Maybe drop the blocks if executor is not able to offload after some timeout - say after 5-10 cycles/retries.

Any thoughts/feedback?

prakharjain09 committed:
minor changes

SparkQA commented:
**[Test build #119763 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119763/testReport)** for PR 27864 at commit [`622e1ba`](https://github.com/apache/spark/commit/622e1ba22aa2b143480282ade1e0b8cff0067d42).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24492/
Test PASSed.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -412,6 +412,14 @@ package object config {
 
+  private[spark] val STORAGE_DECOMMISSION_ENABLED =
+    ConfigBuilder("spark.storage.decommission.enabled")
+      .booleanConf.createWithDefault(false)
```
Please add `version` and split the line like the following.
```scala
private[spark] val STORAGE_DECOMMISSION_ENABLED =
     ConfigBuilder("spark.storage.decommission.enabled")
       .version("3.1.0")
       .booleanConf
       .createWithDefault(false)
```

prakharjain09 replied:
done.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -412,6 +412,14 @@ package object config {
+
+  private[spark] val STORAGE_DECOMMISSION_MAX_REPLICATION_FAILURE =
+    ConfigBuilder("spark.storage.decommission.maxReplicationFailures")
+      .intConf.createWithDefault(3)
```
ditto.

prakharjain09 replied:
done.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1829,7 +1895,45 @@ private[spark] class BlockManager(
     data.dispose()
   }
 
+  class BlockManagerDecommissionManager {
```
Could you add a class description?

dongjoon-hyun replied:
Also, `private[spark] class`.

prakharjain09 replied:
Updated.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1761,6 +1776,57 @@ private[spark] class BlockManager(
+    }
+
+    // Maximum number of storage replication failure which replicateBlock can handle
+    // before giving up for one block
```
1. We had better have this message at the configuration declaration part.
2. We may need to revise the conf name to explicitly show this clause, `before giving up for one block`.

prakharjain09 replied:
@dongjoon-hyun Renamed to "spark.storage.decommission.maxReplicationFailuresPerBlock". Any suggestions on 2) ?

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1761,6 +1776,57 @@ private[spark] class BlockManager(
   }
 
+  def decommissionBlockManager(): Unit = {
+    if (!blockManagerDecommissioning) {
```
Shall we have a `logDebug` when it's already decommissioning (`blockManagerDecommissioning = true`)?

prakharjain09 replied:
done.

prakharjain09 replied:
done.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMaster.scala:
```diff
@@ -43,6 +43,19 @@ class BlockManagerMaster(
+
+  /** Get Replication Info for all the RDD blocks stored in given blockManagerId */
+  def getReplicateInfoForRDDBlocks(
+      blockManagerId: BlockManagerId): Seq[ReplicateBlock] = {
```
It seems that you don't need to break in this case.
```scala
def getReplicateInfoForRDDBlocks(blockManagerId: BlockManagerId): Seq[ReplicateBlock] = {
```

prakharjain09 replied:
fixed.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMaster.scala:
```diff
@@ -43,6 +43,19 @@ class BlockManagerMaster(
+  def getReplicateInfoForRDDBlocks(
+      blockManagerId: BlockManagerId): Seq[ReplicateBlock] = {
+    driverEndpoint.askSync[Seq[ReplicateBlock]](
+      GetReplicateInfoForRDDBlocks(blockManagerId))
```
ditto.
```scala
driverEndpoint.askSync[Seq[ReplicateBlock]](GetReplicateInfoForRDDBlocks(blockManagerId))
```

prakharjain09 replied:
fixed.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMaster.scala:
```diff
@@ -43,6 +43,19 @@ class BlockManagerMaster(
+      GetReplicateInfoForRDDBlocks(blockManagerId))
+  }
+
+
```
redundant empty line.

prakharjain09 replied:
removed.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:
```diff
@@ -65,6 +65,10 @@ class BlockManagerMasterEndpoint(
+  // Set of block managers which are decommissioning
+  private val decommissioningBlockManagerSet = new mutable.HashSet[BlockManagerId]
+
+
```
redundant empty line.

prakharjain09 replied:
removed.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMessages.scala:
```diff
@@ -36,6 +36,9 @@ private[spark] object BlockManagerMessages {
     extends ToBlockManagerSlave
 
+  case object DecommissionBlockManager
+    extends ToBlockManagerSlave
```
Let's make this one-liner since it's simple.

prakharjain09 replied:
Updated.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1665,9 +1678,11 @@ private[spark] class BlockManager(
     }
 
-    logDebug(s"block $blockId replicated to ${peersReplicatedTo.mkString(", ")}")
+    logInfo(s"block $blockId replicated to ${peersReplicatedTo.mkString(", ")}")
```
I believe we should not change this.

prakharjain09 replied:
Reverted.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:
```diff
@@ -299,6 +312,42 @@ class BlockManagerMasterEndpoint(
+   */
+  def decommissionBlockManagers(
+      blockManagerIds: Seq[BlockManagerId]): Future[Seq[Unit]] = {
+
```
redundant empty line.

prakharjain09 replied:
done.

SparkQA commented:
**[Test build #119763 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119763/testReport)** for PR 27864 at commit [`622e1ba`](https://github.com/apache/spark/commit/622e1ba22aa2b143480282ade1e0b8cff0067d42).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/119763/
Test PASSed.

prakharjain09 committed:
review comments addressed

SparkQA commented:
**[Test build #119774 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119774/testReport)** for PR 27864 at commit [`d792092`](https://github.com/apache/spark/commit/d792092569a3f63da6aa303f962974b547197078).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24504/
Test PASSed.

SparkQA commented:
**[Test build #119774 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119774/testReport)** for PR 27864 at commit [`d792092`](https://github.com/apache/spark/commit/d792092569a3f63da6aa303f962974b547197078).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/119774/
Test PASSed.

prakharjain09 commented:
@Dooyoung-Hwang Thanks for the review! I have addressed most of the review comments.

@cloud-fan @holdenk Please review the changes and provide feedback.

holdenk commented:
So maybe not in this PR, but rather than kill tasks which have the chance of succeeding what about if cached with replication level 2? The logic might get a bit funky so I think a separate follow up PR there would be better, but this way we get the best of both worlds.

holdenk commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -412,6 +412,21 @@ package object config {
+  private[spark] val STORAGE_DECOMMISSION_MAX_REPLICATION_FAILURE_PER_BLOCK =
+    ConfigBuilder("spark.storage.decommission.maxReplicationFailuresPerBlock")
+      .doc("Maximum number of failures to tolerate for offloading " +
+        "one block in single decommission cache blocks iteration")
```
I think we could work on the wording here but not critical.

holdenk commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -568,7 +581,7 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   private[spark] def decommissionExecutor(executorId: String): Unit = {
     if (driverEndpoint != null) {
-      logInfo("Propegating executor decommission to driver.")
+      logInfo("Propogating executor decommission to driver.")
```
My bad on misspelling this originally, but I think the correct one is "Propagating"?

prakharjain09 replied:
fixed.

holdenk commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -652,7 +665,7 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   /**
    * Update the cluster manager on our scheduling needs. Three bits of information are included
    * to help it make decisions.
-   * @param resourceProfileToNumExecutors The total number of executors we'd like to have per
```
thanks for the cleanup +1

holdenk commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1281,6 +1285,9 @@ private[spark] class BlockManager(
 
     require(blockId != null, "BlockId is null")
     require(level != null && level.isValid, "StorageLevel is null or invalid")
+    if (blockManagerDecommissioning && blockId.isRDD) {
```
I don't think we should reject the block?

prakharjain09 replied:
Added overall design to explain why I went ahead with this -  https://github.com/apache/spark/pull/27864#issuecomment-605599508 .

holdenk replied:
Still not sure about this part of the design. I'd like to see a test to make sure this won't cause task failures.

prakharjain09 replied:
@holdenk Fixed this. Removed the code for not accepting new RDD cache blocks on Decommissioning executor. Also added a test to show that ongoing tasks which are going to cache data in future succeeds as part of BlockManagerDecommissionSuite.

holdenk commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1560,26 +1567,30 @@ private[spark] class BlockManager(
       existingReplicas: Set[BlockManagerId],
-      maxReplicas: Int): Unit = {
+      maxReplicas: Int,
+      forceFetch: Boolean = true,
```
Why are we making this configurable? And also can we improve the parameter name since this is just controlling force fetching of the peer list.

prakharjain09 replied:
Parameter name changes to forceFetchPeers.

To decommision this block managers, we have to replicate all the cache blocks to some other peers. To replicate n blocks, we are going to call this method n times. This method force fetches the peers each time it is called. So it will end up making n getPeer calls which are not required. So we have made it configurable (by default to true). The decommission flow refreshes the peer list once in beginning and then calls replicateBlock(forceFetchPeers=false) to avoid n calls.

prakharjain09 replied:
forceFetch is not needed to be configurable anymore.

holdenk commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1560,26 +1567,30 @@ private[spark] class BlockManager(
+      maxReplicas: Int,
+      forceFetch: Boolean = true,
+      maxReplicationFailures: Option[Int] = None): Boolean = {
+    var replicatedSuccessfully = true
```
Personally, I don't like var's we could factor this out by using map + headOption + getOrElse but that might not be any cleaner so no pressure on this suggestions.

prakharjain09 replied:
fixed.

holdenk commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1761,6 +1776,59 @@ private[spark] class BlockManager(
+    if (!blockManagerDecommissioning) {
+      logInfo("Starting block manager decommissioning process")
+      blockManagerDecommissioning = true
+      decommissionManager = Some(new BlockManagerDecommissionManager)
```
I guess I'm not sure why were doing this in this manner? It seems like if we did this as a blocking call and returned true when we were done we could move on to a next phase of decommissioning once the blocks were replicated in the driver?

prakharjain09 replied:
@holdenk I have added the overall design for this change. Please review the same and provide your suggestions/feedback.

prakharjain09 replied:
Resolving this as we have another comment thread going on about why a thread is needed instead of blocking call.

holdenk commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1829,7 +1897,49 @@ private[spark] class BlockManager(
+   */
+  private class BlockManagerDecommissionManager {
+    @volatile private var stopped = false
+    private val cacheReplicationThread = new Thread {
```
Same comment about the thread up above, I'm not sure why we need to be in a separate thread (but maybe I've missed something).

prakharjain09 replied:
Tried to explain why we have created a thread here - https://github.com/apache/spark/pull/27864#issuecomment-605599508. Basically this BlockManager will keep on retrying offloading of cache blocks every 30 seconds.

holdenk replied:
Makes sense. I guess we can just (later on) send a message back once all the blocks are done migrating.

prakharjain09 replied:
When all the blocks are done migrating - Driver will automatically know. BlockManagerMaster is continuously receiving BlockUpdates. So when all blocks are offloaded - Dynamic allocation will atomatically know that this executor doesn't have cache data and If other criteria are met (like executorIdleTimeout etc), then the executor will automatically be removed by ExecutorMonitor class of DynamicAllocation.

holdenk reviewed (commented):
Thank you so much for working on this, I'm really you glad you picked it up. I have some questions about the design and some small places for improvement, but really excited.

holdenk commented:
Also, how would you feel about adding an integration test with K8s? We could use a local object which errors on recompute and verify that decommissioning and a second count doesn't trigger it.

prakharjain09 commented:
> Thank you so much for working on this, I'm really you glad you picked it up. I have some questions about the design and some small places for improvement, but really excited.

@holdenk Thanks for the review. My bad, I should have added the overall design initially itself.

Current overall design:

1) CoarseGrainedSchedulerBackend receives a signal to decommissionExecutor. On receiving the signal, it do 2 things - Stop assigning new tasks (SPARK-20628), Send another message to BlockManagerMasterEndpoint (via BlockManagerMaster) to decommission the corresponding BlockManager.

2) BlockManagerMasterEndpoint receives "DecommissionBlockManagers" message. On receiving this, 
 - it updates the "decommissioningBlockManagerSet". This set contains all BMs which are undergoing decommissioning. This set is maintained to giver the correct peerList to the active BMs. Active BMs keep on asking BlockManagerMasterEndpoint for peer list. We should only give active BM's as part of peer list. 

 - it sends a message to the corresponding BlockManager to start the decommissioning process.


3) BlockManager on worker (say BM-x) receives the "DecommissionBlockManager" message. Now it will take following 2 actions on receiving the message - 
- The BM-x will stop accepting new RDD Cache blocks. Since this block manager is in decommissioning state - so it shouldn't store any new cache data (this is similar to DataNode decommissioning in HDFS). If it accepts new blocks - then those blocks also needs to be offloaded.
- BM-x will start BlockManagerDecommissionManager thread. This thread will try to offload any cache blocks on this BM every 30 seconds (or configured time). In the first attempt - it might be possible that all cache blocks are not offloaded because of limited space on other BMs, so it can retry offloading after 30 seconds and so on... The available cache space in the application can keep on changing, because of dynamic allocation/or user can explicitly uncache some other RDD. So we can do multiple attempts to offload RDD cache blocks from BM-x. Steps performed in single attempt: 
a) Ask for replication info of all the cache blocks on BM-x with BlockManagerMasterEndpoint. This will let BM-x know which block is already replicated at what all places - So that we can avoid replicating at those places.
b) For each block - try to replicate it to peers. Drop the block if block is successfully replicated to one of the peer, else keep the block.

prakharjain09 committed:
review comments

SparkQA commented:
**[Test build #120554 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120554/testReport)** for PR 27864 at commit [`b9906c2`](https://github.com/apache/spark/commit/b9906c2d48bd34ec7e181c12d5175ba2dbcd2adc).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25260/
Test PASSed.

SparkQA commented:
**[Test build #120554 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120554/testReport)** for PR 27864 at commit [`b9906c2`](https://github.com/apache/spark/commit/b9906c2d48bd34ec7e181c12d5175ba2dbcd2adc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120554/
Test PASSed.

prakharjain09 committed:
fix comment

SparkQA commented:
**[Test build #120727 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120727/testReport)** for PR 27864 at commit [`076dd67`](https://github.com/apache/spark/commit/076dd671d9a9ac3c8d5926279f54517b8b39426a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25425/
Test PASSed.

SparkQA commented:
**[Test build #120727 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120727/testReport)** for PR 27864 at commit [`076dd67`](https://github.com/apache/spark/commit/076dd671d9a9ac3c8d5926279f54517b8b39426a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120727/
Test FAILed.

holdenk commented:
So I'm curious on the behaviour of rejecting a block, for replication sure but if there was a task still running locally and it cached a block I think the right behaviour would be to accept the cache block then try to replicate it.

holdenk commented:
The documentation failure seems to probably be unrelated (JAR resolve failure). Jenkins retest this please.

SparkQA commented:
**[Test build #120737 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120737/testReport)** for PR 27864 at commit [`076dd67`](https://github.com/apache/spark/commit/076dd671d9a9ac3c8d5926279f54517b8b39426a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25435/
Test PASSed.

holdenk commented on core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:
```diff
@@ -1706,6 +1706,89 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
+    assert(master.getLocations(block3).size === 1)
+  }
+
+  test("test decommission block manager should not be part of peers") {
```
I'd like to see a test where we decomission a block manager and execute a task that attempts to cache the result. I'm worried the current structure might cause that task to fail unnecessarily.

prakharjain09 replied:
Yes. Updated the code. Added new test to specifically test this scenario in BlockManagerDecommissionSuite.

SparkQA commented:
**[Test build #120737 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120737/testReport)** for PR 27864 at commit [`076dd67`](https://github.com/apache/spark/commit/076dd671d9a9ac3c8d5926279f54517b8b39426a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120737/
Test PASSed.

prakharjain09 committed:
accept new blocks even on a decommissiong block manager

SparkQA commented:
**[Test build #120875 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120875/testReport)** for PR 27864 at commit [`d12dbff`](https://github.com/apache/spark/commit/d12dbffa3402fbfdb7d7b38e6b1a9cc94a518630).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25567/
Test PASSed.

prakharjain09 commented:
> Also, how would you feel about adding an integration test with K8s? We could use a local object which errors on recompute and verify that decommissioning and a second count doesn't trigger it.

Added a UT in BlockManagerDecommissionSuite to exactly check that recomputation of RDD doesn't trigger recomputation.

SparkQA commented:
**[Test build #120875 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120875/testReport)** for PR 27864 at commit [`d12dbff`](https://github.com/apache/spark/commit/d12dbffa3402fbfdb7d7b38e6b1a9cc94a518630).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120875/
Test FAILed.

holdenk commented:
Jenkins retest this please.

SparkQA commented:
**[Test build #120885 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120885/testReport)** for PR 27864 at commit [`d12dbff`](https://github.com/apache/spark/commit/d12dbffa3402fbfdb7d7b38e6b1a9cc94a518630).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25577/
Test PASSed.

SparkQA commented:
**[Test build #120885 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120885/testReport)** for PR 27864 at commit [`d12dbff`](https://github.com/apache/spark/commit/d12dbffa3402fbfdb7d7b38e6b1a9cc94a518630).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120885/
Test FAILed.

prakharjain09 committed:
Merge remote-tracking branch 'apache/master' into SPARK-20732-rddcache-1

SparkQA commented:
**[Test build #120912 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120912/testReport)** for PR 27864 at commit [`4c67660`](https://github.com/apache/spark/commit/4c676601ab5e8e73bedf9c686291a51a9ea9780b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25604/
Test PASSed.

SparkQA commented:
**[Test build #120912 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120912/testReport)** for PR 27864 at commit [`4c67660`](https://github.com/apache/spark/commit/4c676601ab5e8e73bedf9c686291a51a9ea9780b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120912/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #120930 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120930/testReport)** for PR 27864 at commit [`4c67660`](https://github.com/apache/spark/commit/4c676601ab5e8e73bedf9c686291a51a9ea9780b).

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -438,6 +438,19 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+
+        if (conf.get(STORAGE_DECOMMISSION_ENABLED)) {
+          try {
+            logInfo(s"Starting decommissioning block manager corresponding to " +
```
nit. `s"` -> `"`.

prakharjain09 replied:
done.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25622/
Test PASSed.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -438,6 +438,19 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+            scheduler.sc.env.blockManager.master.decommissionBlockManagers(Seq(executorId))
+          } catch {
+            case e: Exception =>
+              logError(s"Unexpected error during block manager " +
```
`s"` -> `"`

prakharjain09 replied:
done.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1761,6 +1775,58 @@ private[spark] class BlockManager(
+
+    if (replicateBlocksInfo.nonEmpty) {
+      logInfo(s"Need to replicate ${replicateBlocksInfo.size} blocks " +
+        s"for block manager decommissioning")
```
`s"` -> `"`

prakharjain09 replied:
done.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1829,7 +1895,50 @@ private[spark] class BlockManager(
+      override def run(): Unit = {
+        while (blockManagerDecommissioning && !stopped) {
+          try {
+            logDebug(s"Attempting to replicate all cached RDD blocks")
```
ditto.

prakharjain09 replied:
done.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1829,7 +1895,50 @@ private[spark] class BlockManager(
+          try {
+            logDebug(s"Attempting to replicate all cached RDD blocks")
+            offloadRddCacheBlocks()
+            logInfo(s"Attempt to replicate all cached blocks done")
```
ditto.

prakharjain09 replied:
done.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1829,7 +1895,50 @@ private[spark] class BlockManager(
+            logDebug(s"Attempting to replicate all cached RDD blocks")
+            offloadRddCacheBlocks()
+            logInfo(s"Attempt to replicate all cached blocks done")
+            val sleepInterval = if (Utils.isTesting) 100 else 30000
```
To avoid magic number, shall we have a conf or a constant variable for this, at least, `30000`?

prakharjain09 replied:
done. Added a config which is overridden in tests.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1829,7 +1895,50 @@ private[spark] class BlockManager(
+            Thread.sleep(sleepInterval)
+          } catch {
+            case _: InterruptedException =>
+            // no-op
```
indentation.

prakharjain09 replied:
fixed.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1829,7 +1895,50 @@ private[spark] class BlockManager(
+    cacheReplicationThread.setDaemon(true)
+    cacheReplicationThread.setName("cache-replication-thread")
+
+    def start(): Unit = {
```
Shall we have a `logInfo` to match with line 1933, `logInfo("Stopping cache replication thread")`?

prakharjain09 replied:
done.

prakharjain09 replied:
added logInfo.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:
```diff
@@ -153,6 +156,14 @@ class BlockManagerMasterEndpoint(
 
+    case DecommissionBlockManagers(executorIds) =>
+      val bmIds = executorIds.flatMap(blockManagerIdByExecutor.get)
+      decommissionBlockManagers(bmIds)
```
It seems that we don't need to declare `bmIds`.
```
-      val bmIds = executorIds.flatMap(blockManagerIdByExecutor.get)                                               │
-      decommissionBlockManagers(bmIds)                                                                            │
+      decommissionBlockManagers(executorIds.flatMap(blockManagerIdByExecutor.get))
```

prakharjain09 replied:
refactored.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -413,6 +413,21 @@ package object config {
 
+  private[spark] val STORAGE_DECOMMISSION_ENABLED =
+    ConfigBuilder("spark.storage.decommission.enabled")
+      .doc("Whether to decommission the block manager when decommissioning executor")
```
Could you explain about the meaning of `decommission` here?

Let say, we have Spark Thrift Server. In that case, dynamic allocation wants to decommission all workers. Then, what this PR can provide for zero-worker situation?

prakharjain09 replied:
@dongjoon-hyun Thanks for the detailed review.

I couldn't understand your question with respect to dynamic allocation completely (as per my understanding dynamic allocation doesn't initiates decommissioning). This change works together with SPARK-20628 [PR](https://github.com/apache/spark/pull/26440). After SPARK-20628, Spark scheduler starts receiving decommissionExecutor message (because of AWS spot loss, Azure low-priority VMs etc). In such case, we should offload cache data from that executor to other active executors so that we don't eventually loose those cache blocks. This is done in a best effort way i.e. move cache blocks if space is available on active executors. If space is not available on other active executors, then keep the cache blocks.

Next step after this change is to initiate decommissioning from DynamicAllocation. Let say minExecutor=2, maxExecutor=50. A spark application is running at max 50 executors and all of them have little bit of cache data (say 2GB out of total available 5GB capacity). So DynamicAllocation is not able to downscale any of the executors (although they are idle). In future, DynamicAllocation can leverage changes in this PR to do defragmentation of cache data to fewer set of executors so that some of the executors can be freed up.

Hope this answers your question.

holdenk replied:
Since this is focused on executor to executor migration supporting zero executors isn't going to be a viable option here. However since were using the same methods for putting blocks, assuming that we add support for storing these blocks in external storage in the future, this approach should be able to be generalized to use that same mechanism if configured.

dongjoon-hyun reviewed (changes requested):
Thank you for updating, @prakharjain09 . This is a best effort approach, so it's important to be clear on the limitation. Could you elaborate more about this PR's limitation at PR description?

SparkQA commented:
**[Test build #120930 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120930/testReport)** for PR 27864 at commit [`4c67660`](https://github.com/apache/spark/commit/4c676601ab5e8e73bedf9c686291a51a9ea9780b).
 * This patch **fails from timeout after a configured wait of `400m`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120930/
Test FAILed.

prakharjain09 committed:
review comments addressed

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25664/
Test PASSed.

SparkQA commented:
**[Test build #120973 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120973/testReport)** for PR 27864 at commit [`f6b4f7c`](https://github.com/apache/spark/commit/f6b4f7cfbcd0f3d6f2ab3ef0a1d37de97c012957).

prakharjain09 committed:
remove extra space

SparkQA commented:
**[Test build #120975 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120975/testReport)** for PR 27864 at commit [`9c6bdb6`](https://github.com/apache/spark/commit/9c6bdb626d228eb2ebf0d0c82e0a0aa8d0e9ec21).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25666/
Test PASSed.

SparkQA commented:
**[Test build #120975 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120975/testReport)** for PR 27864 at commit [`9c6bdb6`](https://github.com/apache/spark/commit/9c6bdb626d228eb2ebf0d0c82e0a0aa8d0e9ec21).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120975/
Test FAILed.

SparkQA commented:
**[Test build #120973 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120973/testReport)** for PR 27864 at commit [`f6b4f7c`](https://github.com/apache/spark/commit/f6b4f7cfbcd0f3d6f2ab3ef0a1d37de97c012957).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120973/
Test FAILed.

prakharjain09 commented:
> Thank you for updating, @prakharjain09 . This is a best effort approach, so it's important to be clear on the limitation. Could you elaborate more about this PR's limitation at PR description?

Done. Added more info in description.

prakharjain09 commented:
@dongjoon-hyun @holdenk Thanks for the review. I have addressed all the review comments.

From last few days, few UTs are failing and they didn't seem related to my changes. Can it be possible that my tests are causing the other parallely running UTs to fail? Or these are unrelated intermittent failures which should be ignored?

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #121040 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121040/testReport)** for PR 27864 at commit [`9c6bdb6`](https://github.com/apache/spark/commit/9c6bdb626d228eb2ebf0d0c82e0a0aa8d0e9ec21).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25731/
Test PASSed.

holdenk commented:
The UT failures are probably intermittent we should get a passing run first but I wouldn't worry about it at this stage unless the same ones are failing consistently.

SparkQA commented:
**[Test build #121040 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121040/testReport)** for PR 27864 at commit [`9c6bdb6`](https://github.com/apache/spark/commit/9c6bdb626d228eb2ebf0d0c82e0a0aa8d0e9ec21).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121040/
Test FAILed.

prakharjain09 commented:
@holdenk @dongjoon-hyun I have addressed most of the review comments. Can you please review the PR and provide feedback. Thanks!

holdenk commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -438,6 +438,19 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+              logError("Unexpected error during block manager " +
+                s"decommissioning for executor $executorId: ${e.toString}", e)
+          }
+          logInfo(s"Finished decommissioning block manager corresponding to $executorId.")
```
Given that this kicks off an async task I'd used a different word than finished (maybe "Acknowledged"?)

prakharjain09 replied:
done.

SparkQA commented:
**[Test build #121381 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121381/testReport)** for PR 27864 at commit [`9c6bdb6`](https://github.com/apache/spark/commit/9c6bdb626d228eb2ebf0d0c82e0a0aa8d0e9ec21).

holdenk reviewed (commented):
Thanks for continuing to work on this I think it is getting close :)

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26064/
Test PASSed.

SparkQA commented:
**[Test build #121381 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121381/testReport)** for PR 27864 at commit [`9c6bdb6`](https://github.com/apache/spark/commit/9c6bdb626d228eb2ebf0d0c82e0a0aa8d0e9ec21).
 * This patch **fails from timeout after a configured wait of `400m`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121381/
Test FAILed.

holdenk commented:
Jenkins retest this please

dongjoon-hyun commented:
Also, cc @dbtsai

dongjoon-hyun commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26112/
Test PASSed.

SparkQA commented:
**[Test build #121428 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121428/testReport)** for PR 27864 at commit [`9c6bdb6`](https://github.com/apache/spark/commit/9c6bdb626d228eb2ebf0d0c82e0a0aa8d0e9ec21).

SparkQA commented:
**[Test build #121428 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121428/testReport)** for PR 27864 at commit [`9c6bdb6`](https://github.com/apache/spark/commit/9c6bdb626d228eb2ebf0d0c82e0a0aa8d0e9ec21).
 * This patch **fails from timeout after a configured wait of `400m`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121428/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #121458 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121458/testReport)** for PR 27864 at commit [`9c6bdb6`](https://github.com/apache/spark/commit/9c6bdb626d228eb2ebf0d0c82e0a0aa8d0e9ec21).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26142/
Test PASSed.

SparkQA commented:
**[Test build #121458 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121458/testReport)** for PR 27864 at commit [`9c6bdb6`](https://github.com/apache/spark/commit/9c6bdb626d228eb2ebf0d0c82e0a0aa8d0e9ec21).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121458/
Test FAILed.

prakharjain09 committed:
parallely replicate blocks

SparkQA commented:
**[Test build #121678 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121678/testReport)** for PR 27864 at commit [`bb324f9`](https://github.com/apache/spark/commit/bb324f946019b8d700c517cc5eb2f7c11dc70cfc).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26356/
Test PASSed.

prakharjain09 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1761,6 +1775,58 @@ private[spark] class BlockManager(
+    // so that we end up prioritize them over each other
+    val blocksFailedReplication = replicateBlocksInfo.filterNot {
+      case ReplicateBlock(blockId, existingReplicas, maxReplicas) =>
+        val replicatedSuccessfully = replicateBlock(
```
@holdenk I am not sure how this will behave when multiple executors on same host machine are decommissioning. And each one of them is doing it in parallel - may cause some sort of network congestion?
I have replaced code to do it in ThreadPool of size 4. Maybe we should make this configurable? any suggestions?

holdenk replied:
Network congestion is certainly a possibility, I think that for now this strike a good balance between simple code and avoiding hanging all transfers if we have one slow target host. We can revisit this in the future if it turns out we need more control in production environments. Sound good?

SparkQA commented:
**[Test build #121678 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121678/testReport)** for PR 27864 at commit [`bb324f9`](https://github.com/apache/spark/commit/bb324f946019b8d700c517cc5eb2f7c11dc70cfc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121678/
Test FAILed.

holdenk commented:
Jenkins retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26379/
Test PASSed.

SparkQA commented:
**[Test build #121700 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121700/testReport)** for PR 27864 at commit [`bb324f9`](https://github.com/apache/spark/commit/bb324f946019b8d700c517cc5eb2f7c11dc70cfc).

holdenk reviewed (commented):
I think this is really good progress. I'd like to have a follow up JIRA to explore if there is a better way to pick the peers / parallelize block migration but I think that can be a follow on PR.
cc @dongjoon-hyun have your issues been addressed?

dongjoon-hyun commented:
Yes, @holdenk . Please feel free to proceed to merge if you think this is good to you.

holdenk commented:
I'll take another look tomorrow with fresh eyes and make sure, I'm a little too sleepy to merge any code today.

SparkQA commented:
**[Test build #121700 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121700/testReport)** for PR 27864 at commit [`bb324f9`](https://github.com/apache/spark/commit/bb324f946019b8d700c517cc5eb2f7c11dc70cfc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121700/
Test FAILed.

holdenk commented:
Jenkins retest this please

SparkQA commented:
**[Test build #121772: Deflake Build #121700 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121772/testReport)** for PR 27864 at commit [`bb324f9`](https://github.com/apache/spark/commit/bb324f946019b8d700c517cc5eb2f7c11dc70cfc).

SparkQA commented:
**[Test build #121775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121775/testReport)** for PR 27864 at commit [`bb324f9`](https://github.com/apache/spark/commit/bb324f946019b8d700c517cc5eb2f7c11dc70cfc).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26450/
Test PASSed.

holdenk commented:
I've filed SPARK-31555 for the follow up work we've been discussing in this PR.

holdenk reviewed (approved):
Looks good to me.

dongjoon-hyun commented:
Thank you all!

SparkQA commented:
**[Test build #121775 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121775/testReport)** for PR 27864 at commit [`bb324f9`](https://github.com/apache/spark/commit/bb324f946019b8d700c517cc5eb2f7c11dc70cfc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121775/
Test FAILed.

SparkQA commented:
**[Test build #121772: Deflake Build #121700 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121772/testReport)** for PR 27864 at commit [`bb324f9`](https://github.com/apache/spark/commit/bb324f946019b8d700c517cc5eb2f7c11dc70cfc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

dongjoon-hyun commented:
Hi, @holdenk , does this pass the Jenkins? I cannot find a successful run for the last commit, bb324f9 .

dongjoon-hyun commented:
This seems to break the master branch.
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-3.2-hive-2.3/551/

holdenk commented:
There was I think unless I misread the Jenkins bot's comments:
![image](https://user-images.githubusercontent.com/59893/80268323-1030c900-865b-11ea-90a9-731a0c571dfb.png)

dongjoon-hyun commented:
The first two things are `start` signal. The last one is `K8s` integration test suite. And, it doesn't run UT.

holdenk commented:
Ah damn it, I misread the K8s one as the regular Jenkins. It looks like it's the shutdown hook getting in the way, but I'll just revert this in master since it's Friday night and we can revisit the PR without the shutdown hook.

dongjoon-hyun commented:
+1 for reverting. Thank you so much!

holdenk commented:
Thanks for catching it :)

holdenk commented:
Reverted in https://github.com/apache/spark/pull/28337

dongjoon-hyun commented:
Thank you!

prakharjain09 commented:
@dongjoon-hyun @holdenk Thanks a lot for following up on this.

I see 4 test failures https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-3.2-hive-2.3/551/ and all of them seems unrelated:

```
Test Result (4 failures / +4)
org.apache.spark.sql.hive.thriftserver.CliSuite.path command
org.apache.spark.sql.hive.thriftserver.CliSuite.Pad Decimal numbers with trailing zeros to the scale of the column
org.apache.spark.sql.hive.thriftserver.CliSuite.SPARK-30049 Should not complain for quotes in commented lines
org.apache.spark.sql.hive.thriftserver.CliSuite.SPARK-30049 Should not complain for quotes in commented with multi-lines
```

What possibly should be the next step? Is there a way to open this pullrequest again or should I open a new pullrequest where I will try fixing the tests?

> It looks like it's the shutdown hook getting in the way
@holdenk Can you please share more pointers on what might be causing the test failures?

dongjoon-hyun commented:
@prakharjain09 . Please create the same PR once more. Jenkins will validate your PR.

dongjoon-hyun commented:
In this PR, there is no evidence for the last commit to pass the all unit tests.
